luavm/src/lib/instructions/functions.rs

```diff
@@ -15,80 +16,174 @@ pub fn closure(vm: &mut Vm, instr: u32) -> Result<(), LuaError> {
 }
 
 pub fn push(vm: &mut Vm, instr: u32) -> Result<(), LuaError> {
-    // push all the variable arguments of the current function to the stack
-    if second_arg(instr) == 1 {
-        let args_count = vm.closure.args_count();
-        let args_start = vm.closure.args_start();
-        // make sure to skip the arguments which are the actual parameters
-        for i in (args_start + vm.closure.param_count())..(args_start + args_count) {
-            let val = vm.stack[i].clone();
-            vm.stack.push(val);
-        }
-    } else {
-        vm.stack
-            .push(vm.registers[first_arg(instr) as usize].clone());
-    }
+    let val = vm.registers[first_arg(instr) as usize].clone();
+    vm.push(val);
+    let ret_val = vm.closure.ret_vals();
+    vm.closure.set_ret_vals(ret_val + third_arg(instr) as usize);
+    Ok(())
+}
+
+pub fn set_top(vm: &mut Vm, instr: u32) -> Result<(), LuaError> {
+    let closure = vm.registers[first_arg(instr) as usize].get_closure()?;
+    closure.set_args_start(vm.top);
     Ok(())
 }
 
+const MOVR_0_0_1: u32 = make_instr(Opcode::MOVR, 0, 0, 1);
+const MOVR_0_0_2: u32 = make_instr(Opcode::MOVR, 0, 0, 2);
+
 pub fn call(vm: &mut Vm, instr: u32) -> Result<(), LuaError> {
+    // The frame of a function has the following structure:
+    // arg1       <------ closure.args_start()
+    // arg2
+    // ---------- <CALL> happens here, the caller pushes the arguments for the callee
+    //            vm.top is here when <CALL> is processed
+    // saved-reg1
+    // saved-reg2
+    // ---------- callee saves the registers which it is going to clobber
+    // ret-val1
+    // ret-val2
+    // ---------- vm.top is here when the callee is ready to return
+
+    // save the state of the caller
     let old_closure = vm.closure.clone();
+    let old_pc = vm.pc;
     vm.closure = vm.registers[first_arg(instr) as usize].get_closure()?;
-    // push the first `reg_num` registers to the stack, as the function will modify these
-    let reg_num = vm.closure.reg_count();
-    for i in 1..reg_num {
-        vm.stack.push(vm.registers[i].clone());
+    let args_start = vm.closure.args_start();
+    let args_count = vm.top - args_start;
+    vm.closure.set_args_count(args_count);
+    // push the first `reg_count` registers to the stack, as the called function
+    // will modify these
+    let reg_count = vm.closure.reg_count();
```

Comment:
If `reg_count` is only used here, why not put it in the `for` loop (i.e. `for i in 1..vm.closure.reg_count()`)?